lib: remove ERR_INVALID_OPT_VALUE and ERR_INVALID_OPT_VALUE_ENCODING

lundibundi opened the pull request:
This will be a start to generalize all argument validation
errors. As currently we throw ARG/OPT, OUT_OF_RANGE, and other more
specific errors.
The OPT errors didn't bring much to the errors as it's just another
variant of ARG error which is sometimes more confusing (some of our code
used OPT errors to denote just argument validation errors presumably
because of similarity of OPT to 'option' and not 'options-object')
and they don't specify the name of the options object where the invalid
value is located. Much better approach would be to just specify path
to the invalid value in the name of the value as it is done in this PR
(i.e. 'options.format', 'options.publicKey.type' etc)

Also since this decreases a variety of errors we have it'd be easier to
reuse validation code across the codebase.

Refs: https://github.com/nodejs/node/pull/31251
Refs: https://github.com/nodejs/node/pull/34070#discussion_r467251009
Signed-off-by: Denys Otrishko <shishugi@gmail.com>

----------------------------------------
Since this is a semver-major anyway I'm also considering integrating `ERR_OUT_OF_RANGE` into `ERR_INVALID_ARG_VALUE.RangeError` as I don't see much benefit of it having a separate error code since everything will be stated in the message anyways, wdyt?

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->
/cc @nodejs/crypto @nodejs/buffer @nodejs/http2 as most affected systems
/cc @jasnell @tniessen

jasnell commented:
@lundibundi :
> ...don't see much benefit of it having a separate error code since everything will be stated in the message anyways, wdyt?

That's where it gets a bit complicated. The reason we have error codes in the first place is to avoid users from having to rely on the message too much, which is what has led to changes in the error message having to be semver-major changes. I think the move to `ERR_INVALID_ARG_VALUE.RangeError` is fine in this particular case but I think we should supplement that also with additional error properties like `error.detail.min` and `error.detail.max`. Users can then use the combination of error code, type, and detail.

addaleax commented:
> but I think we should supplement that also with additional error properties like `error.detail.min` and `error.detail.max`. Users can then use the combination of error code, type, and detail.

Yeah, I think that’s where we messed up pretty badly when we introduced the error system – we made the errors parametrized, but only exposed those parameters via the message and not programmatically. That pushes people *towards* error message parsing, not away from it.

That being said, for this particular PR, I don’t have a strong opinion about whether we should change the error codes or not. It likely doesn’t matter because all of these are programming errors, not runtime errors (e.g. POSIX basically covers all of these through `EINVAL` – there’s no real point in further distinguishing, because whatever the program just did, it was not supposed to do that).

jasnell commented:
@addaleax:
> Yeah, I think that’s where we messed up pretty badly when we introduced the error system

Without question. At the time I started introducing the error codes, there was fairly strong resistance against doing too much here and I could never get momentum around adding the detail to the error objects. We definitely can and should tho!

BridgeAR commented:
It is possible to add arbitrary properties to the errors, e.g.: https://github.com/nodejs/node/blob/64acae2e0681ba02f99c0f2d3ee9c2058c0f7e46/lib/internal/errors.js#L1376-L1381

BridgeAR commented:
I personally would keep `ERR_OUT_OF_RANGE`. What we do need is a way to group errors together to logical units such as `NetworkErrors`, `InputError` and similar. That would be most useful for users, since it's often not important if it's one exact error code or another as long as it's one of a specific group.

lundibundi commented:
> That's where it gets a bit complicated. The reason we have error codes in the first place is to avoid users from having to rely on the message too much, which is what has led to changes in the error message having to be semver-major changes. I think the move to ERR_INVALID_ARG_VALUE.RangeError is fine in this particular case but I think we should supplement that also with additional error properties like error.detail.min and error.detail.max. Users can then use the combination of error code, type, and detail.

While I agree that error-code separation is important and we definitely should rework the errors we have today to both include the values (like range, allowed values, erroring value etc) and the subsystem (or build new error hierarchy with subsystems though considering this is JS it may be hander to work with than just `this.type === 'network'` etc) as noted by @BridgeAR.

But for errors such as these (related to argument validation, which is basically EINVAL in POSIX), I don't see much use of having separate error codes since they are just programming errors and should not be programmatically\runtime handled (or am I missing some use-case?). This is exactly why I proposed removing `ERR_OUT_OF_RANGE` in favor of `ERR_INVALID_ARG_VALUE.RangeError`.

> Without question. At the time I started introducing the error codes, there was fairly strong resistance against doing too much here and I could never get momentum around adding the detail to the error objects. We definitely can and should tho!

Yep, when I tried to make some changes to validation or errors in other PRs they didn't get much traction either, unfortunately. Should we start some sort of WG or at least add a group with people interested?

lundibundi commented:
Rebased and fixed conflicts, ready for review.
Also, instead of removing `ERR_INVALID_OPT_VALUE` and `ERR_INVALID_OPT_VALUE_ENCODING` from docs I moved them to `Legacy errors` in `errors.md`.

The removal/replacement of `ERR_OUT_OF_RANGE` with custom properties on RangeError will be done in a separate PR after this lands to have a proper discussion there.

/cc @jasnell @Trott @addaleax @BridgeAR

BridgeAR commented on doc/api/buffer.md:
```diff
@@ -288,6 +288,10 @@ It can be constructed in a variety of ways.
 added: v5.10.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/#####
```
```suggestion
    pr-url: https://github.com/nodejs/node/pull/34682
```

BridgeAR commented on doc/api/buffer.md:
```diff
@@ -353,6 +357,10 @@ A `TypeError` will be thrown if `size` is not a number.
 added: v5.10.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/#####
```
```suggestion
    pr-url: https://github.com/nodejs/node/pull/34682
```

BridgeAR commented on doc/api/buffer.md:
```diff
@@ -401,12 +409,17 @@ additional performance that [`Buffer.allocUnsafe()`][] provides.
 added: v5.12.0
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/#####
```
```suggestion
    pr-url: https://github.com/nodejs/node/pull/34682
```

BridgeAR commented on lib/internal/errors.js:
```diff
@@ -1080,7 +1080,8 @@ E('ERR_INVALID_ARG_VALUE', (name, value, reason = 'is invalid') => {
   }
-  return `The argument '${name}' ${reason}. Received ${inspected}`;
+  const type = name.includes('.') ? 'property' : 'argument';
+  return `The ${type} '${name}' ${reason}. Received ${inspected}`;
```
While changing this, we might also want to change the order to:

```suggestion
  return `The '${name}' ${type} ${reason}. Received ${inspected}`;
```

This seems nicer to read to me.

lundibundi replied:
Looks better to me as well, let's ping a few more people since this incurs lots of test changes
@Trott @jasnell @addaleax @mcollina wdyt?

lundibundi commented:
@Trott Is this ready to land or am I missing something? (commit-queue should be able to handle this PR)

mhdawson commented:
What's the thought on how much ecosystem breakage might be caused in the ecosystem? With my TSC hat on that is my biggest concern.

lundibundi commented:
@mhdawson Personally I'd not expect much breakage because of this. It's an argument validation error change that is unlikely to happen in production and furthermore I don't think anyone would check for the error code in that case. The only "breakage" I can see is an invalid error in some of the old code examples which is not that important and perhaps some CLI tools or wrappers (though I don't know any) that can print helpful messages based on error codes.

We have a CITGM run for this but I don't think it's representative as checking for argument-validation error code is very much an edge case.

Trott commented:
> @Trott Is this ready to land or am I missing something? (commit-queue should be able to handle this PR)

Looks ready to land as far as I can tell.

Trott commented on doc/api/errors.md:
```diff
@@ -2391,6 +2381,16 @@ Used when an invalid character is found in an HTTP response status message
 A given index was out of the accepted range (e.g. negative offsets).
 
+<a id="ERR_INVALID_OPT_VALUE"></a>
+### `ERR_INVALID_OPT_VALUE`
```
For legacy errors, I think we usually add YAML metadata indicating when an error was added and removed. But I don't know how/when it is added.

Trott replied:
(This can probably land without it and a follow-on PR can add it, but if anyone knows what kind of YAML placeholders to add here, let's get 'em added.)

lundibundi replied:
Found: https://github.com/nodejs/node/pull/11317 and https://github.com/nodejs/node/pull/11300, those will be v9 and v8 respectively.
I'll add them to also re-run the GitHub Actions run.

lundibundi commented:
ping @Trott @jasnell @BridgeAR @mcollina this needs another review before it lands since there were changes to errors.md file to include version history in yaml (latest fixup!) as noted by @Trott.

Trott commented:
> ping @Trott @jasnell @BridgeAR @mcollina this needs another review before it lands since there were changes to errors.md file to include version history in yaml (latest fixup!) as noted by @Trott.

The fix-up commit looks good to me.

Trott reviewed (approved):
Fixup commit LGTM. Rubber-stamp on the rest of it based on other's who have already reviewed it.
